feat: optimize url construction

baldurh opened the pull request:
## Description

This PR tries to address the issues presented in the issues #328 and #387 by optimizing the code used to construct urls.

- Removes usage of `jsuri` from `constructUrlFromParams` and replaces it with custom code
- Optimizes how `buildSrc` constructs `srcSet` by reusing the constructed url.

I’m aware that introducing custom code for managing query parameters is a potential risk.

Related issue: #328, #387

**Flamegraph before**
![image](https://user-images.githubusercontent.com/1823617/60345433-8ff0f700-99a8-11e9-9785-5c982c5a5d1f.png)


**Flamegraph after**
![image](https://user-images.githubusercontent.com/1823617/60345455-9da67c80-99a8-11e9-94b0-a9f407baa733.png)

**Flamegraph final** (including changes to fixed size url calculations)
![image](https://user-images.githubusercontent.com/1823617/60424208-5c4be200-9bdf-11e9-9ca0-573ef315e751.png)

jayeb commented on src/constructUrl.js:
```diff
@@ -106,12 +105,18 @@ var DEFAULT_OPTIONS = Object.freeze({
+    ...params
+  };
+  return `${url}?${Object.entries(allParams)
+    .map(([a, b]) => `${a}=${encodeURIComponent(b)}`)
```
This would double-encode any parameters that came in pre-encoded in `src`, right? Perhaps those parameters should be naively decoded during the map step on L109.

baldurh replied:
Yes. Good point. I didn’t think of that.

jayeb commented on src/react-imgix.jsx:
```diff
@@ -9,6 +9,7 @@ import constructUrl from "./constructUrl";
 import { deprecatePropsHOC, ShouldComponentUpdateHOC } from "./HOCs";
 
 import { warning, shallowEqual, compose, config, CONSTANTS } from "./common";
+import { isGenericTypeAnnotation } from "@babel/types";
```
How does this work, exactly? Importing this type annotation affects the output file, even though the imported object is never referenced?

baldurh replied:
Whoops! VS code must have added this automatically when I’ve mistakenly used some suggested option. I’ll remove this and rebase. This slipped passed me. Nice catch ✌️

baldurh commented:
@jayeb I’m happy to help 🙂
I’ve addressed your comments. Please take a look

frederickfogerty commented on src/react-imgix.jsx:
```diff
@@ -124,25 +124,28 @@ function buildSrc({
+        showARWarning = true;
+      }
+
       const buildSrcSetPair = targetWidth => {
```
Optional: this function could be moved out into the module scope, and the variables used passed in as parameters.

frederickfogerty replied:
If this is done, I'm also curious as to what effect memoization could have on the performance of this function. My guess is that it wouldn't improve performance... but miracles can happen.

frederickfogerty commented on src/react-imgix.jsx:
```diff
@@ -124,25 +124,28 @@ function buildSrc({
           aspectRatioDecimal > 0
         ) {
-          urlParams.height = Math.ceil(targetWidth / aspectRatioDecimal);
+          url = `${url}&h=${Math.ceil(targetWidth / aspectRatioDecimal)}`;
```
Optional: maybe there's more performance to be gained from doing this concatenation (along with L139) in one go with L148? (the template string in the return statement). Something like 

```js
let h;

// ...

h = Math.ceil(targetWidth / aspectRatioDecimal);

// ...

return `${url}&h={h}&w=${targetWidth} ${targetWidth}w`; 
```

baldurh replied:
In that case we might get `&h=&w=...`. Is that acceptable?

baldurh replied:
Never mind. I changed the behaviour a little bit 🙂

frederickfogerty commented on src/constructUrl.js:
```diff
@@ -5,7 +5,6 @@ Licensed under the Apache License 2.0, seen https://github.com/coursera/react-im
 Minor syntax modifications have been made
 */
 
-var Uri = require("jsuri");
```
Now that jsuri is only used in the tests, this can moved to a devDep 👍

baldurh commented:
@frederickfogerty Thanks for your suggestions 🙂
I made some changes. Mainly, I moved `buildSrcSetPair` out of `buildSrc` like you suggested and I changed the function so it does as little work as possible for each srcSet pair. It basically returns different build functions based on whether there is a fixed height, aspect ratio or no height given. This way we are pretty much "memoizing" since the only work being done for each srcSet pair is just appending the width.

baldurh commented:
Basic time measurements (on my laptop)
Running `buildSrc` 10,000 times
**Current version**: 3698.878ms
**This branch**: 876.097ms

frederickfogerty commented:
@baldurh Prettier doesn't seem happy with the code formatting here - this can be fixed by running `npm run pretty` and committing the changes. Usually Prettier would auto-fix this for us but it doesn't work with outsider PRs 😬.

If you have any more issues with Prettier, tag me and I'll jump in to help.

baldurh commented:
@frederickfogerty yeah I noticed that and I already took care to run it. Now when I run it there are no changes locally. So I guess the configuration is a bit off maybe? Also I can't see what the problem is when I click the thing so I can't fix it manually either 🙈

frederickfogerty commented:
@baldurh I ran `npm i && npm run pretty` and there was a change... I just can't quite work out what it is. Potentially a line ending, as it doesn't look like whitespace.

Here's a `git diff`:

![image](https://user-images.githubusercontent.com/615334/60468052-389f9080-9c0d-11e9-845c-86742a891f2e.png)

frederickfogerty commented:
@baldurh I went and edited the file directly, and Prettier seems happy now. Just waiting on a 👍 from @jayeb and we should be good to go!

baldurh commented:
@frederickfogerty that's so weird. I don't know why it didn't work for me. Anyway, thanks for your help 👌🏻

frederickfogerty commented:
@baldurh Usually the culprit is an out-of-date prettier installation (global or local) 👍. But of course there are other reasons as well. Glad I could help out!

baldurh commented:
@jayeb Any chance you could take another look? We have a pending PR in our project we don’t want to merge until this is added 😬

sherwinski commented:
Hey @baldurh 👋 it looks like @jayeb will be unavailable to look at this until Friday if you're able to wait until then.

baldurh commented:
@sherwinski sure no problem 😊

baldurh commented:
> This is so clean now, I love it. Thanks for all the hard work @baldurh.

My pleasure. This is a fun project 🙂

sherwinski commented:
@baldurh these changes are now available in [v8.6.2](https://github.com/imgix/react-imgix/releases/tag/v8.6.2) 🎉 
Thanks again for all your help!
